frontend-react/docs/test-conventions.md

```diff
@@ -0,0 +1,504 @@
+
+### Tests should read like a manual for source code
+
+Tests are often one of the first places to look when dealing with familiar code because they offer insight into how specific pieces function in isolation.  Because of this, we should strive to make our tests read like a manual for the related source code: "given this input, I should expect this output" or more practically, "given these conditions, I should expect this behavior."  Most concepts in programming should be able to be boiled down into inputs and outputs, and tests are no different in that regard.
+
+```tsx
+function sum(...numbers: number[]) {
+    return numbers.reduce((total, value) => total + value, 0);
+}
+
+// bad - what does it mean to function correctly?
+describe("sum", () => {
+    test("functions correctly", () => {
+        expect(sum()).toEqual(0);
+
+        expect(sum(1)).toEqual(1);
+        expect(sum(1, 2)).toEqual(3);
+        expect(sum(1, 2, 3)).toEqual(6);
+        expect(sum(1, 10, 100, 1000)).toEqual(1111);
+    });
+});
+
+// good - use descriptions to explain what's actually happening
+describe("sum", () => {
+    describe("with no arguments", () => {
+        test("returns 0", () => {
+            expect(sum()).toEqual(0);
+        });
+    });
+
+    describe("with arguments", () => {
+        test("returns the sum of all the arguments", () => {
+            expect(sum(1)).toEqual(1);
+            expect(sum(1, 2)).toEqual(3);
+            expect(sum(1, 2, 3)).toEqual(6);
+            expect(sum(1, 10, 100, 1000)).toEqual(1111);
+        });
+    });
+})
+```
+
+> Why?
+>
+> Having clearer and more sensible descriptions calls attention to what the purpose of the test actually is, which makes it easier to follow along with the underlying logic or to group similarly intentioned tests together.
+
+
+### Tests should check behavior, not implementation
+
+Especially in the world of front-end code, it's especially important to test our code from a user's perspective -- this means our tests should be prioritizing checking behavior ("what happens") rather than implementation details ("how it happens").
+
+```tsx
+function List({ items }: { items: string[] }) {
+    return (
+        <ul className="bg-light-gray">
+            { items.slice(0, 10).map((item) => (
+                <li className="bg-white" key={item}>{item}</li>
+            )) }
+        </ul>
+    );
+}
+
+// bad - don't test markup or classes, which are more likely to change over time
+test("renders ul with bg-light-gray", () => {
+    expect(ulNode).toHaveClass("bg-light-gray");
+});
+
+// good - check what the user would see
+describe("when there are ten or more items", () => {
+    test("renders ten items", () => {
+        expect(liNodes).toHaveLength(10);
+    });
+});
```

Comment:
It's true that we should avoid testing CSS classes, but markup is inevitable. Perhaps not something as specific as `expect <span>` or something, but we're always looking for DOM nodes in a way. I wonder if we can lay best practices there? We could even get as explicit as using `data` attributes for nodes we're using and expecting for tests?

Reply:
Sure, that's a good point; we're always going to be testing markup implicitly, but it's rarely important to test markup _explicitly_.  Will clarify!

Reply:
Actually, RTL has a great doc about this already: https://testing-library.com/docs/queries/about/#priority

Maybe we could just link to this as our guideline on this point?

Reply:
Yeah! Love that idea. Especially that `getByTestId` both exists AND should be used as a last resort.